src/modm/driver/display/st7586s_protocol.hpp

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2021, Tomasz Wasilczyk
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+#ifndef MODM_ST7586S_HPP
+#error "Do not include this file directly. Include st7586s.hpp instead."
+#endif
```

Comment:
Not precisely :). Ifndef protects from including this file before the main header, pragma protects from including it after.

Reply:
I'm a student (as well) - never have seen such a construct. Any guess @salkinium ?

Reply:
I think primarily because modm keeps implementation in "cpp" header files included at the end of the actual header files and not just in source files. It's already an unique setup.

Reply:
> modm doesn't seem to stick to a single name for this stuff. Interface definitions are scattered around files named: `_defines`, `_transport`, `_register`, `_definitions`, `_data`, `_constants`. `register` is only used by ssd1306 driver actually.
> 
> @salkinium, what's your take? I don't think this library should stick to old conventions, but that's not my call at the end of the day :)

Confirm, we need a convention for future drivers so this unifies. I like `_defines` the most cause it takes the classic culture from ANSI-C. Anything else is already to tight.

Reply:
Yea most of the implementations are templates, therefore these `*_impl.hpp`.

Reply:
It's mainly just historic convention to have just one header file which pulls in everything that is needed. We can remove this construct completely and instead just `#pragma once` and then `#include "st7586s.hpp"`, which will also make IDE code completion work in this file, since it's pulling in the correct definitions.

(Ok, this file is dependency free, so you don't need the `#include "st7586s.hpp"` part, but in the `st7586s.hpp_impl.hpp` file that can be used).

Reply:
Genius. Thanks!

I didn't add the include to the *protocol details* ( ;) ) file, since it doesn't depend on the main one indeed and there may be valid uses for it separately.